kube-runtime/src/events.rs

```diff
@@ -184,74 +224,136 @@ impl Recorder {
+            reporting_controller: Some(self.reporter.controller.clone()),
+            reporting_instance: Some(
+                self.reporter
+                    .instance
+                    .clone()
+                    .unwrap_or_else(|| self.reporter.controller.clone()),
+            ),
+            series: None,
+            type_: match ev.type_ {
+                EventType::Normal => Some("Normal".into()),
+                EventType::Warning => Some("Warning".into()),
+            },
+            related: ev.secondary.clone(),
         }
     }
 
     /// Publish a new Kubernetes' event.
     ///
     /// # Access control
     ///
-    /// The event object is created in the same namespace of the [`ObjectReference`]
-    /// you specified in [`Recorder::new`].
+    /// The event object is created in the same namespace of the [`ObjectReference`].
     /// Make sure that your controller has `create` permissions in the required namespaces
     /// for the `event` resource in the API group `events.k8s.io`.
     ///
     /// # Errors
     ///
     /// Returns an [`Error`](`kube_client::Error`) if the event is rejected by Kubernetes.
-    pub async fn publish(&self, ev: Event) -> Result<(), kube_client::Error> {
-        // See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#event-v1-events-k8s-io
-        // for more detail on the fields
-        // and what's expected: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#event-v125
-        self.events
-            .create(&PostParams::default(), &K8sEvent {
-                action: Some(ev.action),
-                reason: Some(ev.reason),
-                deprecated_count: None,
-                deprecated_first_timestamp: None,
-                deprecated_last_timestamp: None,
-                deprecated_source: None,
-                event_time: Some(MicroTime(Utc::now())),
-                regarding: Some(self.reference.clone()),
-                note: ev.note.map(Into::into),
-                metadata: ObjectMeta {
-                    namespace: self.reference.namespace.clone(),
-                    generate_name: Some(format!("{}-", self.reporter.controller)),
-                    ..Default::default()
-                },
-                reporting_controller: Some(self.reporter.controller.clone()),
-                reporting_instance: Some(
-                    self.reporter
-                        .instance
-                        .clone()
-                        .unwrap_or_else(|| self.reporter.controller.clone()),
-                ),
-                series: None,
-                type_: match ev.type_ {
-                    EventType::Normal => Some("Normal".into()),
-                    EventType::Warning => Some("Warning".into()),
-                },
-                related: ev.secondary,
-            })
-            .await?;
+    pub async fn publish(&self, ev: Event, reference: &ObjectReference) -> Result<(), kube_client::Error> {
```

Comment:
have put a comment elsewhere about the particular api breaking part (which can maybe be avoided). otherwise, i think it's fine to replace the behaviour, since it's kind of just upgrading it under the hood.